idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/editor/hints/HintingSyntax.kt

```diff
@@ -14,114 +14,374 @@ import com.intellij.lang.parameterInfo.ParameterInfoHandlerWithTabActionSupport
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.ClassifierDescriptor
+import org.jetbrains.kotlin.renderer.ClassifierNamePolicy
+import org.jetbrains.kotlin.renderer.DescriptorRenderer
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+import org.jetbrains.kotlin.descriptors.CallableDescriptor
+import org.jetbrains.kotlin.psi.KtDestructuringDeclarationEntry
+import org.jetbrains.kotlin.psi.KtFunction
+import org.jetbrains.kotlin.psi.KtIfExpression
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtTryExpression
+import org.jetbrains.kotlin.psi.KtWhenExpression
 
+/**
+ * Hint's are generally used with TypeInferenceAlgorithms.
+ * [HintingSyntax] provides means to connect those algorithms to the ide for specified element's.
+ */
 interface HintingSyntax {
-  fun IdeMetaPlugin.addExpressionTypeProvider(
-    informationHint: (element: KtExpression) -> String,
-    expressionAt: (elementAt: PsiElement) -> MutableList<KtExpression>,
-    errorHint: String
+  //TODO: consider if PsiElement.selectionTextRangeOnTextEditor from the following example is universal
+
+  /**
+   * registers a [ExpressionTypeProvider] for [KtExpression]s.
+   * Use this function to add an improved TypeInferenceAlgorithm for Kotlin to the ide.
+   * [informationHint] and [expressionAt] will be executed as the user types.
+   * The following examples is a minimal version of the
+   * [org.jetbrains.kotlin.idea.codeInsight.KotlinExpressionTypeProvider] and targets [KtFunction]s, who are expressed as expressions.
+   *
+   * ```kotlin:ank:playground
+   * import arrow.meta.Plugin
+   * import arrow.meta.ide.IdeMetaPlugin
+   * import arrow.meta.invoke
+   * import com.intellij.openapi.editor.ex.util.EditorUtil
+   * import com.intellij.openapi.fileEditor.FileEditorManager
+   * import com.intellij.openapi.fileEditor.TextEditor
+   * import com.intellij.openapi.util.TextRange
+   * import com.intellij.openapi.vfs.VirtualFile
+   * import com.intellij.psi.PsiElement
+   * import org.jetbrains.kotlin.descriptors.CallableDescriptor
+   * import org.jetbrains.kotlin.psi.KtFunction
+   * import org.jetbrains.kotlin.psi.psiUtil.parentsWithSelf
+   * import org.jetbrains.kotlin.renderer.DescriptorRenderer
+   * import org.jetbrains.kotlin.renderer.RenderingFormat
+   * import org.jetbrains.kotlin.resolve.BindingContext
+   * import org.jetbrains.kotlin.types.KotlinType
+   * import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+   *
+   * //sampleStart
+   * val IdeMetaPlugin.expressionHints: Plugin
+   *  get() = "Hints for KtExpressions" {
+   *   meta(
+   *    addExpressionTypeProviderForKotlin(
+   *     informationHint = { expr ->
+   *      get(BindingContext.DECLARATION_TO_DESCRIPTOR, expr).safeAs<CallableDescriptor>()?.returnType
+   *       ?.let { result: KotlinType ->
+   *         typeRenderer.renderType(result)
+   *        } ?: "Type is unknown"
+   *    },
+   *    expressionAt = {
```

Comment:
I have to admit that I don't understand this example. I've no experience with the extension points of the Kotlin plugin.
The docs of Kotlin's expressionAt is
```
  /**
   * Returns the list of all possible targets at specified position.
   */
  @NotNull
  public abstract List<T> getExpressionsAt(@NotNull PsiElement elementAt);
```

This example isn't using the input element passed to `expressionAt`, but is using the current editor instead. If this is correct, then as a user reading this I'd like to know why the input parameter isn't used.

Is it certain that this function is always called from an editor? Accessing the editor might be a bit expensive, depending on how often this is called (I didn't check).

Reply:
It is you may have overwritten it but `it` in the first line is `elementAt` in your example. 
in `Meta` most of these functions are flatten for a straight forward implementation.

Reply:
Thanks. 
So this is `KotlinExpressionTypeProvider.getExpressionAt(elementAt: PsiElement)` with a few modifications. `KotlinExpressionTypeProvider`  also uses the passed PsiElement only to get the editor.

Reply:
It doesn't use the editor. it solely passes valid KtExpressions, which are at the beginning of the textRange.
It collects all valid KtFunctions and returns the one which starts at the same `startoffSet`.

Reply:
I meant that it only uses `selectionTextRangeOnTextEditor()` on the input element, which retrieves the selected text range from the editor, where the containing file is edited.

Reply:
exactly :)